Allow configurable attribute name on `#has_secure_password`

choncou opened the pull request:
### Summary

This addition will now allow configuring an attribute name for the
existing `#has_secure_password` method. This can be useful when one would
like to store some secure field as a digest, just like a password.

This also creates a method `#authenticate_XXX` where XXX is
the configured attribute name on `#has_secure_password`. `#authenticate`
is now an alias to this method when the attribute name is the default
'password'.

Commit:
This addition will now allow configuring an attribute name for the
existing `#has_secure_password`. This can be useful when one would
like to store some secure field as a digest, just like a password.

The method still defaults to `password`. It now also allows using the
same `#authenticate` method which now accepts a second argument for
specifying the attribute to be authenticated, or will default to 'password`.

A new method is also added for generating a new token for an attribute by
calling `#regenerate_XXXX` where `XXXX` is the attribute name.

matthewd commented on activemodel/lib/active_model/secure_password.rb:
```diff
@@ -91,36 +135,9 @@ module InstanceMethodsOnActivation
-
-      def password_confirmation=(unencrypted_password)
-        @password_confirmation = unencrypted_password
+      def authenticate(unencrypted_password, attribute = :password)
```
Let's call this `authenticate_#{attribute}` (with an alias when attribute=='password')

matthewd commented on activemodel/lib/active_model/secure_password.rb:
```diff
@@ -61,23 +68,60 @@ def has_secure_password(options = {})
+        #   user.password = nil
+        #   user.password_digest # => nil
+        #   user.password = 'mUc3m00RsqyRe'
+        #   user.password_digest # => "$2a$10$4LEA7r4YmNHtvlAvHhsYAeZmk/xeUVtMTYqwIvYY76EW5GUqDiP4."
```
We're losing this documentation. Maybe we should push it up to the module on a fake `password=` attribute... or otherwise, we should merge it into the `has_secure_password` documentation. Either way, it's no good here.

matthewd commented on activemodel/lib/active_model/secure_password.rb:
```diff
@@ -61,23 +68,60 @@ def has_secure_password(options = {})
+
+        define_method("regenerate_#{attribute}") do
+          self.send("#{attribute}=", self.class.generate_unique_secure_token)
+        end
```
We [still](https://www.reddit.com/r/rails/comments/55f60p/code_reviewfeedback_a_more_secure_version_of_has/d8aa26e) don't need this.

Commit:
Remove method for regenerating a token, and update `#authenticate`.

This change now creates a method `#authenticate_XXX` where XXX is
the configured attribute name on `#has_secure_password`. `#authenticate`
is now an alias to this method when the attribute name is the default
'password'

choncou commented:
@matthewd Any further feedback/changes I should look into?

NomNomCameron commented:
Any update on this? Would be super nice to be able to have this for other columns on models that require a digest hash of something other than `password`

choncou commented:
@matthewd do you have any direction/feedback in order for us to get this through?

@NomNomCameron This has been silent for the past year, I ended up just adding the functionality into the my own projects manually by adding this updated */active_model/secure_password.rb*

alexstophel commented:
:+1: for getting this through. I'd love this feature.

shideneyu commented:
Is there any plans on merging this PR ? People really need it.
